File: modules/flyright-document-import/ios/FlyRightDocumentImportModule.swift
import ImageIO
import ExpoModulesCore
import PDFKit
import Vision

/// iOS half of the shared-document reader — see ../index.ts for the contract.
///
/// Intake needs no native code here: a PDF shared via "Copy to FlyRight" is
/// delivered as a file:// URL to `application(_:open:options:)`, which
/// expo-linking already forwards and src/app/+native-intent.ts rewrites into
/// the import route. This module only reads the file: PDFKit for the text of
/// each page, Vision for every barcode drawn on it (PDF417 on printed passes
/// and Amadeus receipts, Aztec/QR on mobile passes).
///
/// `readImage` is the same job for a picture instead of a PDF — a screenshot
/// of a mobile pass, a photo of a paper one — so an in-app upload lands in
/// the same import screen. Vision reads the barcode; when the image has none
/// (a screenshot of a confirmation email) its text recogniser stands in for
/// PDFKit's page text.
public class FlyRightDocumentImportModule: Module {
  public func definition() -> ModuleDefinition {
    Name("FlyRightDocumentImport")

    Events("onDocumentShared")

    // Cold-start shares reach JS through the linking URL on iOS; nothing to hand over.
    Function("consumePendingDocument") { () -> [String: Any]? in
      return nil
    }

    AsyncFunction("readPdf") { (uri: URL, maxPages: Int, promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          promise.resolve(try Self.read(url: uri, maxPages: maxPages))
        } catch {
          promise.reject(error)
        }
      }
    }

    AsyncFunction("readImage") { (uri: URL, promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          promise.resolve(try Self.readImage(url: uri))
        } catch {
          promise.reject(error)
        }
      }
    }
  }

  /// Pages render at 3x their PDF points (216 dpi): dense enough for Vision to
  /// resolve the narrow bars of a PDF417 stripe, small enough (~1800x2400 for
  /// Letter) to stay well inside memory on every supported iPhone.
  private static let renderScale: CGFloat = 3

  /// Second pass for a page that came back empty. A stripe printed small (a
  /// Finnair e-ticket's is 1.4 inches wide) or a page scanned at low quality
  /// is a few pixels per module at 216 dpi; twice that reads it. Only ever
  /// runs when the cheap pass found nothing, so the usual document still
  /// costs one render.
  private static let retryRenderScale: CGFloat = 6

  private static func validateImport(_ url: URL) throws {
    let root = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0].appendingPathComponent("document-imports").resolvingSymlinksInPath().standardizedFileURL
    let file = url.resolvingSymlinksInPath().standardizedFileURL
    let values = try file.resourceValues(forKeys: [.isRegularFileKey, .fileSizeKey])
    guard file.isFileURL, file.deletingLastPathComponent().path == root.path, values.isRegularFile == true,
      let size = values.fileSize, size > 0, size <= 20 * 1024 * 1024 else { throw DocumentUnreadableException("Invalid document") }
  }

  private static func read(url: URL, maxPages: Int) throws -> [String: Any] {
    try validateImport(url)
    // No-op for Inbox copies; needed if the file were opened in place.
    let scoped = url.startAccessingSecurityScopedResource()
    defer {
      if scoped { url.stopAccessingSecurityScopedResource() }
    }

    guard let document = PDFDocument(url: url) else {
      throw DocumentUnreadableException(url.lastPathComponent)
    }
    if document.isLocked {
      throw DocumentLockedException()
    }

    var pages: [[String: Any]] = []
    for index in 0..<min(document.pageCount, min(max(maxPages, 1), 8)) {
      guard let page = document.page(at: index) else { continue }
      pages.append([
        "text": rowOrderedText(on: page),
        "barcodes": barcodes(on: page),
      ])
    }
    return ["pageCount": document.pageCount, "pages": pages]
  }

  /// A page's text in reading order, the same shape the OCR path produces and
  /// the same order PDFBox's sortByPosition gives on Android.
  ///
  /// `PDFPage.string` is content-stream order, not visual order, and a table
  /// drawn column by column comes out interleaved. On a Finnair e-ticket the
  /// operating airline lands on a line of its own while the "Operated by"
  /// label runs straight into the *marketing* column, so the text reads
  /// "Operated by FINNAIR" about a leg British Airways flies — PDFKit itself
  /// reports label and far value as one contiguous range. The parser matches
  /// labels to values by proximity, so it has no way to see past that.
  ///
  /// Geometry is the only thing that still knows the truth, so rows are
  /// rebuilt from it: each character is placed by its own bounds, banded into
  /// rows by a typical glyph height, ordered left to right, and any gap wider
  /// than a word space — including the single very wide space PDFKit puts
  /// between two columns — becomes the double space the itinerary parser
  /// reads as a column break. Falls back to `page.string` if the page carries
  /// no measurable glyph.
  ///
  /// `characterBounds(at:)` is deliberately not used: its indices drift out of
  /// step with `string` on this document. A one-character selection is both
  /// index-correct and cheap (~2 ms for a page).
  private static func rowOrderedText(on page: PDFPage) -> String {
    let raw = String((page.string ?? "").prefix(100_000))
    let ns = raw as NSString
    guard ns.length > 0 else { return raw }

    var boxes: [CGRect] = []
    boxes.reserveCapacity(ns.length)
    for index in 0..<ns.length {
      let bounds = page.selection(for: NSRange(location: index, length: 1))?.bounds(for: page)
      if let bounds, [bounds.minX, bounds.maxX, bounds.minY, bounds.maxY, bounds.width, bounds.height].allSatisfy({ $0.isFinite && abs($0) < 10_000_000 }) {
        boxes.append(bounds)
      } else { boxes.append(.null) }
    }
    let heights = boxes.filter { !$0.isNull && $0.height > 0 }.map(\.height).sorted()
    guard !heights.isEmpty, let top = boxes.filter({ !$0.isNull }).map(\.maxY).max() else {
      return raw
    }
    let band = max(heights[heights.count / 2], 1)
    // Wider than any word space at this text size: a column break.
    let columnGap = band * 0.8

    var glyphs: [(row: Int, minX: CGFloat, maxX: CGFloat, text: String, isColumnBreak: Bool)] = []
    glyphs.reserveCapacity(ns.length)
    for index in 0..<ns.length {
      let box = boxes[index]
      if box.isNull || box.isInfinite { continue }
      let character = ns.substring(with: NSRange(location: index, length: 1))
      if character == "\n" || character == "\r" { continue }
      let blank = character.trimmingCharacters(in: .whitespaces).isEmpty
      // PDF's origin is bottom-left; flip so the row index grows downward.
      let row = Int(((top - box.midY) / band).rounded(.down))
      glyphs.append((row, box.minX, box.maxX, character, blank && box.width > columnGap))
    }
    guard !glyphs.isEmpty else { return raw }
    glyphs.sort { $0.row != $1.row ? $0.row < $1.row : $0.minX < $1.minX }

    var lines: [String] = []
    var line = ""
    var row = glyphs[0].row
    var previousMaxX: CGFloat?
    for glyph in glyphs {
      if glyph.row != row {
        lines.append(line)
        line = ""
        row = glyph.row
        previousMaxX = nil
      }
      let gapped = previousMaxX.map { glyph.minX - $0 > columnGap } ?? false
      if (glyph.isColumnBreak || gapped) && !line.isEmpty && !line.hasSuffix("  ") {
        line += "  "
      }
      if !glyph.isColumnBreak { line += glyph.text }
      previousMaxX = glyph.maxX
    }
    lines.append(line)
    return lines
      .map { $0.trimmingCharacters(in: .whitespaces) }
      .filter { !$0.isEmpty }
      .joined(separator: "\n")
  }

  private static func barcodes(on page: PDFPage) -> [String] {
    let found = barcodes(on: page, scale: renderScale)
    if !found.isEmpty { return found }
    return barcodes(on: page, scale: retryRenderScale)
  }

  private static func barcodes(on page: PDFPage, scale renderScale: CGFloat) -> [String] {
    let bounds = page.bounds(for: .mediaBox)
    let size = CGSize(width: bounds.width * renderScale, height: bounds.height * renderScale)
    guard bounds.origin.x.isFinite, bounds.origin.y.isFinite, size.width.isFinite, size.height.isFinite, size.width > 0, size.height > 0, size.width * size.height <= 40_000_000 else { return [] }
    // Draw the page ourselves at exactly renderScale: PDFPage.thumbnail(of:)
    // multiplies by the screen scale on iOS (a 9x page on a 3x device) and
    // Vision rejects or mis-reads the result.
    let format = UIGraphicsImageRendererFormat()
    format.scale = 1
    format.opaque = true
    let image = UIGraphicsImageRenderer(size: size, format: format).image { context in
      UIColor.white.setFill()
      context.fill(CGRect(origin: .zero, size: size))
      let cg = context.cgContext
      // PDF space is bottom-left origin; flip into UIKit's top-left.
      cg.translateBy(x: 0, y: size.height)
      cg.scaleBy(x: renderScale, y: -renderScale)
      cg.translateBy(x: -bounds.origin.x, y: -bounds.origin.y)
      page.draw(with: .mediaBox, to: cg)
    }
    guard let cgImage = image.cgImage else { return [] }
    return barcodes(in: cgImage)
  }

  private static func barcodes(in cgImage: CGImage) -> [String] {
    let request = VNDetectBarcodesRequest()
    request.symbologies = [.pdf417, .qr, .aztec, .dataMatrix]
    #if targetEnvironment(simulator)
    // The simulator has no inference context for Vision's ML detector
    // ("Could not create inference context"). The classic revision runs
    // there and reads the larger stripes (1 of the receipt's 2), enough for
    // development builds to exercise the barcode path. Devices keep the
    // default revision, which decodes everything (verified with the same
    // PDFKit + Vision calls on macOS).
    //
    // Some stripes are invisible to the classic detector at any render
    // scale — a Finnair e-ticket's 1.4-inch PDF417 reads on the default
    // revision at 3x and never on revision 1, even at 8x (measured). A
    // document that imports on a phone can therefore be refused on the
    // simulator; that is the simulator, not the reader.
    if VNDetectBarcodesRequest.supportedRevisions.contains(VNDetectBarcodesRequestRevision1) {
      request.revision = VNDetectBarcodesRequestRevision1
    }
    #endif
    let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
    do {
      try handler.perform([request])
    } catch {
      NSLog("[FlyRightDocumentImport] barcode detection failed on %dx%d page: %@", cgImage.width, cgImage.height, error.localizedDescription)
      return []
    }

    var seen = Set<String>()
    var payloads: [String] = []
    for observation in request.results ?? [] {
      guard let payload = observation.payloadStringValue, !payload.isEmpty else { continue }
      if seen.insert(payload).inserted { payloads.append(payload) }
    }
    return payloads
  }

  // -- images ---------------------------------------------------------------

  /// Vision's detectors work on the pixels they are given, so an upload is
  /// read the same way a rendered PDF page is — one "page" out, so the pure
  /// extractor downstream needs no notion of images at all.
  private static func readImage(url: URL) throws -> [String: Any] {
    try validateImport(url)
    let scoped = url.startAccessingSecurityScopedResource()
    defer {
      if scoped { url.stopAccessingSecurityScopedResource() }
    }

    guard let source = CGImageSourceCreateWithURL(url as CFURL, nil),
      let properties = CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [CFString: Any],
      let width = properties[kCGImagePropertyPixelWidth] as? Double,
      let height = properties[kCGImagePropertyPixelHeight] as? Double,
      width > 0, height > 0, width * height <= 40_000_000,
      let data = try? Data(contentsOf: url), let image = UIImage(data: data),
      let cgImage = downscaled(image)
    else {
      throw ImageUnreadableException(url.lastPathComponent)
    }
    return [
      "pageCount": 1,
      "pages": [["text": text(in: cgImage), "barcodes": barcodes(in: cgImage)]],
    ]
  }

  /// A 12-megapixel camera photo is more pixels than either detector needs
  /// and enough to matter for peak memory; anything already smaller is left
  /// exactly as it is, since a barcode loses bars to resampling. Redrawing
  /// also normalises the EXIF orientation a phone photo carries, which the
  /// raw CGImage would not.
  private static let maxImageEdge: CGFloat = 3200

  private static func downscaled(_ image: UIImage) -> CGImage? {
    let size = image.size
    guard size.width > 0, size.height > 0 else { return nil }
    let scale = min(1, maxImageEdge / max(size.width, size.height))
    let target = CGSize(width: (size.width * scale).rounded(), height: (size.height * scale).rounded())
    let format = UIGraphicsImageRendererFormat()
    format.scale = 1
    format.opaque = true
    let drawn = UIGraphicsImageRenderer(size: target, format: format).image { context in
      UIColor.white.setFill()
      context.fill(CGRect(origin: .zero, size: target))
      image.draw(in: CGRect(origin: .zero, size: target))
    }
    return drawn.cgImage
  }

  /// Text as printed, row by row: the extractor reads proximity inside the
  /// string, so an itinerary table has to come out in reading order or a
  /// leg's airports end up far from its flight number. Lines are banded into
  /// rows by a typical line height and ordered left to right within a row —
  /// the same order PDFBox's sortByPosition gives on Android. Language
  /// correction stays off; it "fixes" flight numbers and airport codes into
  /// words.
  private static func text(in cgImage: CGImage) -> String {
    let request = VNRecognizeTextRequest()
    request.recognitionLevel = .accurate
    request.usesLanguageCorrection = false
    let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
    do {
      try handler.perform([request])
    } catch {
      NSLog("[FlyRightDocumentImport] text recognition failed: %@", error.localizedDescription)
      return ""
    }
    let observations = request.results ?? []
    let heights = observations.map { $0.boundingBox.height }.sorted()
    let band = max(heights.isEmpty ? 0.01 : heights[heights.count / 2], 0.001)
    let lines = observations
      .compactMap { observation -> (row: Int, x: CGFloat, text: String)? in
        guard let string = observation.topCandidates(1).first?.string else { return nil }
        let box = observation.boundingBox
        // Vision's origin is bottom-left; flip so the row index grows downward.
        return (Int(((1 - box.midY) / band).rounded(.down)), box.minX, string)
      }
      .sorted { $0.row != $1.row ? $0.row < $1.row : $0.x < $1.x }
      .map(\.text)
    return lines.joined(separator: "\n")
  }
}

private final class DocumentUnreadableException: GenericException<String> {
  override var reason: String {
    "'\(param)' could not be opened as a PDF."
  }
}

private final class ImageUnreadableException: GenericException<String> {
  override var reason: String {
    "'\(param)' could not be opened as an image."
  }
}

private final class DocumentLockedException: Exception {
  override var reason: String {
    "This PDF is password-protected."
  }
}

File: modules/flyright-document-import/android/src/main/java/expo/modules/flyrightdocumentimport/FlyRightDocumentImportModule.kt
package expo.modules.flyrightdocumentimport

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.Matrix
import android.graphics.pdf.PdfRenderer
import android.net.Uri
import android.os.Build
import android.os.ParcelFileDescriptor
import android.provider.OpenableColumns
import androidx.exifinterface.media.ExifInterface
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.barcode.BarcodeScanner
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.google.zxing.BarcodeFormat
import com.google.zxing.BinaryBitmap
import com.google.zxing.DecodeHintType
import com.google.zxing.MultiFormatReader
import com.google.zxing.RGBLuminanceSource
import com.google.zxing.common.HybridBinarizer
import com.google.zxing.multi.GenericMultipleBarcodeReader
import com.tom_roush.pdfbox.android.PDFBoxResourceLoader
import com.tom_roush.pdfbox.pdmodel.PDDocument
import com.tom_roush.pdfbox.text.PDFTextStripper
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.util.UUID

/** Android half of the shared-document reader — see ../index.ts for the
 * contract.
 *
 * Intake: "Share → FlyRight" on a PDF starts (or re-enters, launchMode
 * singleTask) MainActivity with an ACTION_SEND intent whose EXTRA_STREAM is a
 * content:// URI the sender granted us to read only while this activity
 * lives. Neither React Native's Linking nor expo-router see that intent, so
 * the module copies the stream into the cache dir the moment it arrives and
 * passes the private file:// path to JS — as an event when the app is already
 * running (OnNewIntent), or held as `pending` for JS to collect after a cold
 * start. The launch intent is neutralised after collection so a reload of
 * the JS bundle does not import the same file twice.
 *
 * Reading: PdfBox-Android for the text of each page (sorted by position, so
 * table rows come out as lines), Android's PdfRenderer to rasterise pages,
 * and ML Kit to decode every barcode on them.
 *
 * `readImage` is the same job for a picture instead of a PDF — a screenshot
 * of a mobile pass, a photo of a paper one — so an in-app upload lands in the
 * same import screen. The barcode pass is identical once the bitmap exists;
 * ML Kit's text recogniser stands in for PDFBox when the image carries no
 * barcode (a screenshot of a confirmation email). */
class FlyRightDocumentImportModule : Module() {
  private var pending: Map<String, Any?>? = null

  override fun definition() = ModuleDefinition {
    Name("FlyRightDocumentImport")

    Events("onDocumentShared")

    OnNewIntent { intent ->
      capture(intent)?.let { doc ->
        pending = doc
        sendEvent("onDocumentShared", doc)
      }
    }

    Function("consumePendingDocument") {
      val activity = appContext.currentActivity
      val fromLaunch = activity?.let { act ->
        capture(act.intent)?.also { neutralise(act) }
      }
      val result = pending ?: fromLaunch
      pending = null
      result
    }

    // Off the JS thread by construction (AsyncFunction); Tasks.await below
    // must not run on main.
    AsyncFunction("readPdf") { uri: String, maxPages: Int ->
      readPdf(uri, maxPages)
    }

    AsyncFunction("readImage") { uri: String ->
      readImage(uri)
    }
  }

  private val context: Context
    get() = requireNotNull(appContext.reactContext)

  // -- intake ---------------------------------------------------------------

  private fun capture(intent: Intent?): Map<String, Any?>? {
    if (intent == null) return null
    val uri: Uri = when (intent.action) {
      Intent.ACTION_SEND ->
        if (Build.VERSION.SDK_INT >= 33) {
          intent.getParcelableExtra(Intent.EXTRA_STREAM, Uri::class.java)
        } else {
          @Suppress("DEPRECATION")
          intent.getParcelableExtra(Intent.EXTRA_STREAM)
        }
      Intent.ACTION_VIEW -> intent.data
      else -> null
    } ?: return null

    val resolver = context.contentResolver
    val name = displayName(uri)
    val type = intent.type ?: resolver.getType(uri)
    val isPdf = type == "application/pdf" || (name?.lowercase()?.endsWith(".pdf") == true)
    if (!isPdf) return null

    val dir = File(context.cacheDir, "shared-documents").apply { mkdirs() }
    val copy = File(dir, "${UUID.randomUUID()}.pdf")
    try {
      resolver.openInputStream(uri)?.use { input -> copy.outputStream().use { input.copyTo(it) } }
        ?: return null
    } catch (e: Exception) {
      copy.delete()
      return null
    }
    return mapOf("uri" to Uri.fromFile(copy).toString(), "name" to name)
  }

  private fun displayName(uri: Uri): String? {
    if (uri.scheme == "file") return uri.lastPathSegment
    return try {
      context.contentResolver.query(uri, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null)?.use { cursor ->
        if (cursor.moveToFirst()) cursor.getString(0) else null
      }
    } catch (e: Exception) {
      null
    }
  }

  /** The share has been collected: make the launch intent look like a plain
   * launcher start so a JS reload doesn't re-import the same document. */
  private fun neutralise(activity: Activity) {
    val intent = Intent(activity.intent)
    intent.action = Intent.ACTION_MAIN
    intent.removeExtra(Intent.EXTRA_STREAM)
    intent.data = null
    intent.type = null
    activity.intent = intent
  }

  // -- reading --------------------------------------------------------------

  /** 4x the PDF points (288 dpi): ML Kit needs more than Vision does to
   * resolve the narrow bars of a small PDF417 stripe (at 3x it read only one
   * of an Amadeus receipt's two). ~2450x3170 ARGB for a Letter page, ~30 MB,
   * one page at a time. */
  private val renderScale = 4f

  /** Second pass for a page that came back empty — a stripe printed small
   * (a Finnair e-ticket's is 1.4 inches wide) is only a few pixels per
   * module at 4x. Only runs when the cheap pass found nothing. */
  private val retryRenderScale = 8f

  private fun readPdf(uri: String, maxPages: Int): Map<String, Any> {
    val path = Uri.parse(uri).path ?: throw DocumentUnreadableException(uri)
    val file = File(path)
    if (!file.exists()) throw DocumentUnreadableException(uri)

    PDFBoxResourceLoader.init(context)
    val texts = mutableListOf<String>()
    val pageCount: Int
    try {
      PDDocument.load(file).use { doc ->
        if (doc.isEncrypted) throw DocumentLockedException()
        pageCount = doc.numberOfPages
        val stripper = PDFTextStripper().apply { sortByPosition = true }
        for (i in 0 until minOf(pageCount, maxOf(maxPages, 1))) {
          stripper.startPage = i + 1
          stripper.endPage = i + 1
          texts.add(stripper.getText(doc))
        }
      }
    } catch (e: CodedException) {
      throw e
    } catch (e: Exception) {
      throw DocumentUnreadableException(uri)
    }

    val barcodes = decodeBarcodes(file, texts.size)
    val pages = texts.mapIndexed { i, text -> mapOf("text" to text, "barcodes" to barcodes[i]) }
    return mapOf("pageCount" to pageCount, "pages" to pages)
  }

  /** The symbologies travel documents actually carry: PDF417 on printed
   * passes and e-ticket receipts, Aztec/QR on mobile ones. */
  private fun newBarcodeScanner(): BarcodeScanner = BarcodeScanning.getClient(
    BarcodeScannerOptions.Builder()
      .setBarcodeFormats(
        Barcode.FORMAT_PDF417,
        Barcode.FORMAT_QR_CODE,
        Barcode.FORMAT_AZTEC,
        Barcode.FORMAT_DATA_MATRIX,
      )
      .build(),
  )

  private fun decodeBarcodes(file: File, pages: Int): List<List<String>> {
    val scanner = newBarcodeScanner()
    val result = MutableList<List<String>>(pages) { emptyList() }
    try {
      ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY).use { pfd ->
        val renderer = PdfRenderer(pfd)
        try {
          for (i in 0 until minOf(pages, renderer.pageCount)) {
            for (scale in listOf(renderScale, retryRenderScale)) {
              val bitmap = renderer.openPage(i).use { page ->
                Bitmap.createBitmap(
                  (page.width * scale).toInt(),
                  (page.height * scale).toInt(),
                  Bitmap.Config.ARGB_8888,
                ).also {
                  it.eraseColor(Color.WHITE)
                  page.render(it, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)
                }
              }
              try {
                result[i] = payloads(scanner, bitmap)
              } finally {
                bitmap.recycle()
              }
              if (result[i].isNotEmpty()) break
            }
          }
        } finally {
          renderer.close()
        }
      }
    } catch (e: Exception) {
      // Rasterising failed (corrupt page tree, etc.): text alone still imports.
    } finally {
      scanner.close()
    }
    return result
  }

  /** Both decoders over one bitmap: ML Kit first, then ZXing for the small
   * PDF417 stripes it skips. Either failing only costs its own findings. */
  private fun payloads(scanner: BarcodeScanner, bitmap: Bitmap): List<String> {
    val found = LinkedHashSet<String>()
    try {
      // rawValue keeps BCBP's whitespace-significant layout intact;
      // displayValue would "clean" it.
      Tasks.await(scanner.process(InputImage.fromBitmap(bitmap, 0)))
        .mapNotNullTo(found) { it.rawValue?.takeIf(String::isNotEmpty) }
    } catch (e: Exception) {
      // ML Kit failing only means ZXing gets the whole job.
    }
    try {
      found.addAll(zxingDecode(bitmap))
    } catch (e: Exception) {
      // Same: an image that fails to scan simply contributes no barcodes.
    }
    return found.toList()
  }

  // -- images ---------------------------------------------------------------

  /** One "page" out, so the pure extractor downstream needs no notion of
   * images at all. */
  private fun readImage(uri: String): Map<String, Any> {
    val path = Uri.parse(uri).path ?: throw ImageUnreadableException(uri)
    val file = File(path)
    if (!file.exists()) throw ImageUnreadableException(uri)
    val bitmap = loadBitmap(file) ?: throw ImageUnreadableException(uri)
    return try {
      val scanner = newBarcodeScanner()
      val codes = try {
        payloads(scanner, bitmap)
      } finally {
        scanner.close()
      }
      mapOf(
        "pageCount" to 1,
        "pages" to listOf(mapOf("text" to imageText(bitmap), "barcodes" to codes)),
      )
    } finally {
      bitmap.recycle()
    }
  }

  /** A 12-megapixel camera photo is more pixels than either detector needs
   * and enough to matter for peak memory (ARGB, a 24 MP one would be ~96 MB),
   * so the longest edge is capped here — still denser than the ~2450 px the
   * PDF path rasterises and reads PDF417 off reliably. A phone photo's EXIF
   * rotation is applied too: ZXing reads raw pixels and would otherwise see
   * the pass sideways. */
  private val maxImageEdge = 3200

  private fun loadBitmap(file: File): Bitmap? {
    val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    BitmapFactory.decodeFile(file.path, bounds)
    if (bounds.outWidth <= 0 || bounds.outHeight <= 0) return null
    var sample = 1
    while (maxOf(bounds.outWidth, bounds.outHeight) / sample > maxImageEdge) sample *= 2
    val options = BitmapFactory.Options().apply {
      inSampleSize = sample
      inPreferredConfig = Bitmap.Config.ARGB_8888
    }
    val bitmap = BitmapFactory.decodeFile(file.path, options) ?: return null
    val degrees = try {
      when (ExifInterface(file).getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL)) {
        ExifInterface.ORIENTATION_ROTATE_90 -> 90f
        ExifInterface.ORIENTATION_ROTATE_180 -> 180f
        ExifInterface.ORIENTATION_ROTATE_270 -> 270f
        else -> 0f
      }
    } catch (e: Exception) {
      0f
    }
    if (degrees == 0f) return bitmap
    val rotated = Bitmap.createBitmap(
      bitmap, 0, 0, bitmap.width, bitmap.height,
      Matrix().apply { postRotate(degrees) }, true,
    )
    if (rotated != bitmap) bitmap.recycle()
    return rotated
  }

  /** Text as printed, in reading order. `Text.text` would hand back ML Kit's
   * own blocks, which group an itinerary table by COLUMN — every departure
   * airport in one run, far from the flight number it belongs to — and the
   * extractor reads proximity inside the string, so the legs fall apart.
   * Sorting the lines by position instead gives the same top-to-bottom,
   * left-to-right order as PDFBox's sortByPosition and Vision's reader. */
  private fun imageText(bitmap: Bitmap): String {
    val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
    return try {
      val lines = Tasks.await(recognizer.process(InputImage.fromBitmap(bitmap, 0)))
        .textBlocks
        .flatMap { it.lines }
        .mapNotNull { line -> line.boundingBox?.let { line.text to it } }
      // Rows are banded by a typical line height so two cells of the same
      // table row stay together instead of sorting by a few stray pixels.
      val band = (lines.map { it.second.height() }.sorted().getOrNull(lines.size / 2) ?: 1)
        .coerceAtLeast(1)
      lines
        .sortedWith(compareBy({ it.second.centerY() / band }, { it.second.left }))
        .joinToString("\n") { it.first }
    } catch (e: Exception) {
      // No OCR (an unavailable model, an unsupported image): the barcode
      // path still imports.
      ""
    } finally {
      recognizer.close()
    }
  }
}

/** ZXing over the whole page, all codes at once. TRY_HARDER makes the
 * PDF417 detector sweep the image instead of sampling its centre. */
private fun zxingDecode(bitmap: Bitmap): List<String> {
  val pixels = IntArray(bitmap.width * bitmap.height)
  bitmap.getPixels(pixels, 0, bitmap.width, 0, 0, bitmap.width, bitmap.height)
  val source = RGBLuminanceSource(bitmap.width, bitmap.height, pixels)
  val hints = mapOf(
    DecodeHintType.TRY_HARDER to true,
    DecodeHintType.POSSIBLE_FORMATS to listOf(
      BarcodeFormat.PDF_417,
      BarcodeFormat.QR_CODE,
      BarcodeFormat.AZTEC,
      BarcodeFormat.DATA_MATRIX,
    ),
  )
  val reader = GenericMultipleBarcodeReader(MultiFormatReader())
  return try {
    reader.decodeMultiple(BinaryBitmap(HybridBinarizer(source)), hints)
      .mapNotNull { it.text?.takeIf(String::isNotEmpty) }
  } catch (e: com.google.zxing.NotFoundException) {
    emptyList()
  }
}

private class DocumentUnreadableException(uri: String) :
  CodedException("ERR_DOCUMENT_UNREADABLE", "'$uri' could not be opened as a PDF.", null)

private class ImageUnreadableException(uri: String) :
  CodedException("ERR_IMAGE_UNREADABLE", "'$uri' could not be opened as an image.", null)

private class DocumentLockedException :
  CodedException("ERR_DOCUMENT_LOCKED", "This PDF is password-protected.", null)
